### ISSUE TITLE
执行lstm时报Floating point exceptionpython27-gcc482/bin/python conf/trainer_config.conf

### ISSUE THREAD

**Issue by stephenchan0693** [PaddlePaddle/Paddle#7209]
在使用栈式双向LSTM用于文本分类时遇到浮点数溢出问题，尝试使用gradient_clipping_threshold或error_clipping_threshold后仍然报错，具体错误信息如下：
Thu Jan  4 15:02:15 2018[1,4]<stdout>:
Thu Jan  4 15:02:15 2018[1,4]<stdout>:Pass 0, Batch 1300, Cost 0.013177, {'__auc_evaluator_0__': 1.0, 'classification_error_evaluator': 0.009999999776482582}
Thu Jan  4 15:04:19 2018[1,5]<stderr>:Thread [139896152655616] Forwarding __lstmemory_1__, 
Thu Jan  4 15:04:19 2018[1,5]<stderr>:*** Aborted at 1515049459 (unix time) try "date -d @1515049459" if you are using GNU date ***
Thu Jan  4 15:04:19 2018[1,5]<stderr>:PC: @                0x0 (unknown)
Thu Jan  4 15:04:19 2018[1,5]<stderr>:*** SIGFPE (@0x7f3c15db0042) received by PID 25812 (TID 0x7f3c1c7c0700) from PID 366673986; stack trace: ***
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c1c397160 (unknown)
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15db0042 hl_avx_lstm_backward_one_sequence<>()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15daf5eb paddle::LstmCompute::backwardOneSequence<>()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15daf73d paddle::LstmCompute::backwardBatch<>()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15dabd7a paddle::LstmLayer::backwardBatch()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15dac439 paddle::LstmLayer::backward()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15e39baf paddle::NeuralNetwork::backward()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c160551c4 GradientMachine::forwardBackward()
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c15d021d9 _wrap_GradientMachine_forwardBackward
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b4cb9 PyEval_EvalFrameEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6b28 PyEval_EvalCodeEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b5d10 PyEval_EvalFrameEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6b28 PyEval_EvalCodeEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b5d10 PyEval_EvalFrameEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6b28 PyEval_EvalCodeEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b5d10 PyEval_EvalFrameEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6b28 PyEval_EvalCodeEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b5d10 PyEval_EvalFrameEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6b28 PyEval_EvalCodeEx
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4b6c52 PyEval_EvalCode
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4e1c7d PyRun_FileExFlags
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4e3501 PyRun_SimpleFileExFlags
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x4159dd Py_Main
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @     0x7f3c1b8f1bd5 __libc_start_main
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @           0x414b71 (unknown)
Thu Jan  4 15:04:19 2018[1,5]<stderr>:    @                0x0 (unknown)
Thu Jan  4 15:04:20 2018[1,5]<stderr>:train.sh: line 239: 25812 Floating point exceptionpython27-gcc482/bin/python conf/trainer_config.conf

求指点

**Comment by wanghaox** [comment]
能部分下相关的训练和数据的代码吗
并且检查输入数据是否有异常

**Comment by stephenchan0693** [comment]
主要的读取数据和训练代码如下：
fea_map = {}
with open(word_dict_file) as f:
    for line in f:
        arr = line.rstrip('\n').split('\t')
        fea_map[arr[0]] = int(arr[1])

def encode_word(word):
    if fea_map.has_key(word):
        return fea_map[word]
    else:                                                                                                                                                                  
        return fea_map['<unk>']

def cluster_data_reader(file_dir, node_id):
    def data_reader():
        files = os.listdir(file_dir)
        for fi in files:
            with open(file_dir + '/' + fi, "r") as f:
                for line in f:
                    arr = line.strip('\n').split('\t')
                    if len(arr) < 2:
                        continue
                    y = int(arr[0])
                    x = [encode_word(w) for w in arr[1].split(' ') if encode_word(w) >= 0]
                    yield x, y
    return data_reader
def fc_net(dict_dim, class_dim=2):                                                                                                                                         
    """
    dnn network definition

    :param dict_dim: size of word dictionary
    :type input_dim: int
    :params class_dim: number of instance class
    :type class_dim: int
    :params emb_dim: embedding vector dimension
    :type emb_dim: int
    """

    # input layers
    data = paddle.layer.data("word", paddle.data_type.integer_value_sequence(dict_dim))
    lbl = paddle.layer.data("label", paddle.data_type.integer_value(class_dim))

    # embedding layer
    emb = paddle.layer.embedding(input=data, size=64)
    
    lstm_para_attr = paddle.attr.Param(initial_std=0., learning_rate=1e-3)
    
    layer_attr_new = paddle.attr.ExtraLayerAttribute(error_clipping_threshold=100.0)

    bias_attr = paddle.attr.Param(initial_std=0., l2_rate=0.)

    relu = paddle.activation.Relu()
    linear = paddle.activation.Linear()
    hid_dim = 256
    lstm_stack_num = 5

    fc = paddle.layer.mixed(
            size=hid_dim,
            input=paddle.layer.full_matrix_projection(input=emb), 
            act=linear,
            layer_attr=layer_attr_new)
    lstm = paddle.layer.lstmemory(
            input=fc, 
            act=relu, 
            param_attr=lstm_para_attr,
            bias_attr=bias_attr)

    for i in range(2, lstm_stack_num + 1):
        inputs = [paddle.layer.full_matrix_projection(input=fc), paddle.layer.full_matrix_projection(input=lstm)]
        fc = paddle.layer.mixed(
                size=hid_dim,
                input=inputs,
                act=linear,
                bias_attr=bias_attr,
                layer_attr=layer_attr_new)
        lstm = paddle.layer.lstmemory(
                input=fc,
                reverse=(i % 2) == 0,
                act=relu,
                param_attr=lstm_para_attr,
                bias_attr=bias_attr)
        inputs = [fc, lstm]
    
    fc_last = paddle.layer.pooling(input=inputs[0], pooling_type=paddle.pooling.Max())
    lstm_last = paddle.layer.pooling(input=inputs[1], pooling_type=paddle.pooling.Max())

    # output layer
    output = paddle.layer.fc(
        input=[fc_last, lstm_last],
        size=class_dim,
        act=paddle.activation.Softmax(),
        bias_attr=bias_attr,
        param_attr=para_attr)

    cost = paddle.layer.classification_cost(input=output, label=lbl)                                                                                                       

    return cost, output, lbl

**Comment by wanghaox** [comment]
请先检查下输入数据长度是否超过hid_dim/4

**Comment by peterzhang2029** [comment]
Closing due to low activity. Feel free to reopen it.

**Comment by jzh1992** [comment]
后来解决了吗
